onnxruntime/core/framework/partial_executor.h

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#ifdef ENABLE_TRAINING
+#pragma once
+
+#include <vector>
+#include <unordered_map>
+#include "core/common/common.h"
+#include "core/common/status.h"
+#include "core/common/logging/logging.h"
+#include "core/framework/iexecutor.h"
+#include "core/framework/framework_common.h"
+#include "core/framework/ml_value.h"
+#include "core/framework/session_state.h"
+#include "core/graph/graph_viewer.h"
+#include "core/framework/op_kernel_context_internal.h"
+#include "core/framework/sequential_executor.h"
+
+namespace onnxruntime {
+class PartialExecutor : public IExecutor {
+ public:
+  PartialExecutor(PartialGraphExecutionState& state, const bool& terminate_flag = false)
+      : state_{state}, terminate_flag_{terminate_flag} {}
+
+  common::Status Execute(const SessionState& session_state, const std::vector<int>& feed_mlvalue_idxs,
+                         const std::vector<OrtValue>& feeds, const std::vector<int>& fetch_mlvalue_idxs,
+                         std::vector<OrtValue>& fetches,
+                         const std::unordered_map<size_t, CustomAllocator>& fetch_allocators,
+                         const logging::Logger& logger) override;
+
```

Comment:
What's the reason why we need a completely new executor given that appears to have a large amount of code duplication?

Reply:
There's a pretty trivial refactoring of `SequentialExecutor::Execute()` that would make it possible to implement PartialExecutor as a thin wrapper. 

However, since one big aim of this PR is to make it clear that it has no impact on the existing inference path, I'd be inclined to do that refactoring in a follow-up PR, which should keep things easier to review.

Reply:
Agreed.

@KeDengMS FYI.